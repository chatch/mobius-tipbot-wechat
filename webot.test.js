const crypto = require('crypto')
const request = require('supertest')

const webot = require('./webot')
const {WX_TOKEN} = require('./config')

const USER = 'odQ_C0gylEtCpMTORAqxQoLd1Zmg'

const calculateSignature = (timestamp, nonce) => {
  const s = [WX_TOKEN, timestamp, nonce].sort().join('')
  return crypto
    .createHash('sha1')
    .update(s)
    .digest('hex')
}

const signatureQuery = () => {
  const nonce = 1
  const timestamp = Math.floor(Date.now() / 1000)
  return {
    nonce,
    openid: USER,
    signature: calculateSignature(timestamp, nonce),
    timestamp,
  }
}

const xmlMsg = msg =>
  `
<xml>
<ToUserName><![CDATA[gh_97f347ad4987]]></ToUserName>
<FromUserName><![CDATA[${USER}]]></FromUserName>
<CreateTime>1533300000</CreateTime>
<MsgType><![CDATA[text]]></MsgType>
<Content><![CDATA[${msg}]]></Content>
<MsgId>6585322584931228410</MsgId>
</xml>
`.replace(/\n/g, '')

const sendMsg = msg =>
  request(webot)
    .post('/wechat')
    .set('Content-Type', 'text/xml')
    .query(signatureQuery())
    .send(xmlMsg(msg))
    .expect(200)

test('unknown command returns error message', () =>
  sendMsg('/coffee').then(rsp => expect(rsp.text).toMatchSnapshot()))

test('/help command returns help text', () =>
  sendMsg('/help').then(rsp => expect(rsp.text).toMatchSnapshot()))
