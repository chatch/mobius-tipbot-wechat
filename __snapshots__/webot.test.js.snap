// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/help command returns help text 1`] = `
"<xml><ToUserName><![CDATA[odQ_C0gylEtCpMTORAqxQoLd1Zmg]]></ToUserName><FromUserName><![CDATA[gh_97f347ad4987]]></FromUserName><CreateTime>1533288151</CreateTime><MsgType><![CDATA[text]]></MsgType><Content><![CDATA[Hi, odQ_C0gylEtCpMTORAqxQoLd1Zmg, I am Mobius TipBot!

/balance will show your earned tips amount.

/withdraw <address> <amount> will withdraw accumulated tips to your Stellar account. Omit amount to withdraw all the tips at once

If you need to create a wallet see the Mobius StellarTerm Wallet Guide! https://bit.ly/mobi_stg

/create <funding_address> <amount> - create and fund your own Tipping Account to send tips from (no more 1 tip per hour limit!). \`<funding_address>\` - is your current Stellar account that will be used fund your Tipping Account. \`<amount>\` - the amount that Tipping Account will be funded with

/unregister <withdraw_address> - unregister your Tipping Account and withdraw your funds. You will revert to tipping from the default pool and be limited to 1 tip per hour  \`<withdraw_address>\` - Stellar address where your funds will be withdrawn to and Tipping Account will be merged into

/my_address - returns your Tipping Account that was created by /create. If you run out of tokens send more here to continue tipping!
]]></Content></xml>"
`;

exports[`unknown command returns error message 1`] = `
"<xml><ToUserName><![CDATA[odQ_C0gylEtCpMTORAqxQoLd1Zmg]]></ToUserName><FromUserName><![CDATA[gh_97f347ad4987]]></FromUserName><CreateTime>1533288151</CreateTime><MsgType><![CDATA[text]]></MsgType><Content><![CDATA[Command [/coffee] unknown.
Use /help to see a list of commands.]]></Content></xml>"
`;
